### PR TITLE
T/89 Changes emitter for tree model document

### PR DESCRIPTION
**Pull request by pjasiun**
Fixes https://github.com/ckeditor/ckeditor5-core/issues/89.

### CODE REVIEW

File: src/document/operation/changeoperation.js [hunk 1]
```diff
@@ -130,6 +134,12 @@ CKEDITOR.define( [
 					value.node.setAttr( newAttr );
 				}
 			}
+
+			return {
+				range: this.range,
+				oldAttr: oldAttr ? oldAttr : undefined,
```
Comment (scofalik) [inline src/document/operation/changeoperation.js]: Is this really needed? Also I don't think that this should be `undefined` while the operation expected argument is `null` when it is "empty". We should stick with one of them and I propose `null` as for me it passes more data. Then, we should simply return `oldAttr` and `newAttr`. Remember to change docs in `document.Document`.

File: src/document/operation/insertoperation.js [hunk 2]
```diff
@@ -58,6 +64,10 @@ CKEDITOR.define( [
 
 		_execute() {
 			this.position.parent.insertChildren( this.position.offset, this.nodeList );
+
+			return {
+				range: new Range( this.position, Position.createFromPositionAndShift( this.position, this.nodeList.length ) )
```
Comment (scofalik) [inline src/document/operation/insertoperation.js]: In my "OT" pull request I introduced `Range.createFromPositionAndOffset` which does exactly that.

File: src/document/operation/moveoperation.js [hunk 3]
```diff
@@ -121,6 +128,11 @@ CKEDITOR.define( [
 			const removedNodes = sourceElement.removeChildren( sourceOffset, this.howMany );
 
 			targetElement.insertChildren( targetOffset, removedNodes );
+
+			return {
+				sourcePosition: this.sourcePosition,
+				range: new Range( this.targetPosition, Position.createFromPositionAndShift( this.targetPosition, this.howMany ) )
+			};
```
Comment (scofalik) [inline src/document/operation/moveoperation.js]: `Range.createFromPositionAndOffset`

File: src/document/position.js [hunk 4]
```diff
@@ -221,6 +221,25 @@ CKEDITOR.define( [ 'document/rootelement', 'utils', 'ckeditorerror' ], ( RootEle
...
+		 *		 const pos = new Position( [ 1, 2, 8 ], root );
+		 *
+		 *		 Position.createFromPositionAndShift( pos, -3 ) // [ 1, 2, 5 ]
+		 *		 Position.createFromPositionAndShift( pos, 5 )  // [ 1, 2, 13 ]
+		 *
+		 * @param {document.Position} referencePosition Reference position.
+		 * @param {Number} shift Offset shift.
+		 * @returns {document.Position}
+		 */
+		static createFromPositionAndShift( referencePosition, shift ) {
```
Comment (scofalik) [inline src/document/position.js]: We may keep it if we want to, but if it's only use is creating "flat" ranges I'd consider removing it.

File: tests/document/document-change-event.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
+/**
```
Comment (scofalik) [inline tests/document/document-change-event.js]: Create seperate directory `tests/document/document` and put `document.js` and `change.js` (or `event-change.js` or `change-event.js`) there?

**Comment by scofalik** [comment]
I've pushed small commit with minor fixes (mostly typos). Please consider comments I've made.

File: src/document/operation/changeoperation.js (hunk 1 above)
Comment (pjasiun) [inline src/document/operation/changeoperation.js]: Agreed. This is a leftover after the previous idea of having separate events for add, removing and changing attribute.

File: src/document/operation/insertoperation.js (hunk 2 above)
Comment (pjasiun) [inline src/document/operation/insertoperation.js]: Changed. Additional I created: https://github.com/ckeditor/ckeditor5-core/issues/107.

File: src/document/operation/moveoperation.js (hunk 3 above)
Comment (pjasiun) [inline src/document/operation/moveoperation.js]: Done.

File: src/document/position.js (hunk 4 above)
Comment (pjasiun) [inline src/document/position.js]: Removed.

File: tests/document/document-change-event.js (hunk 5 above)
Comment (pjasiun) [inline tests/document/document-change-event.js]: A was also thinking about it. Done.

**Comment by scofalik** [comment]
I've added one commit with removed dependency and merged the branch.

**Pull request merged**